File: MIDTERM_LAB/CustomerSegmentation.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class CustomerSegmentation {

    // Segment thresholds (in thousands of units = total sales)
    private static final double PLATINUM_MIN = 100000;
    private static final double GOLD_MIN     = 50000;
    private static final double SILVER_MIN   = 10000;

    private List<VideoGame> games;

    // Segment buckets: publisher name -> total sales
    private Map<String, Double> publisherSales;

    private List<String> platinum;
    private List<String> gold;
    private List<String> silver;
    private List<String> bronze;

    public CustomerSegmentation(List<VideoGame> games) {
        this.games          = games;
        this.publisherSales = new HashMap<>();
        this.platinum       = new ArrayList<>();
        this.gold           = new ArrayList<>();
        this.silver         = new ArrayList<>();
        this.bronze         = new ArrayList<>();
    }

    public void analyze() {
        // Step 1: Aggregate total sales per publisher
        for (VideoGame g : games) {
            String publisher = g.getPublisher();
            double sales     = g.getTotalSales();
            if (!publisher.equals("N/A") && sales > 0) {
                if (publisherSales.containsKey(publisher)) {
                    publisherSales.put(publisher, publisherSales.get(publisher) + sales);
                } else {
                    publisherSales.put(publisher, sales);
                }
            }
        }

        // Step 2: Segment each publisher
        for (Map.Entry<String, Double> entry : publisherSales.entrySet()) {
            String publisher = entry.getKey();
            double sales     = entry.getValue();

            if (sales > PLATINUM_MIN) {
                platinum.add(publisher);
            } else if (sales >= GOLD_MIN) {
                gold.add(publisher);
            } else if (sales >= SILVER_MIN) {
                silver.add(publisher);
            } else {
                bronze.add(publisher);
            }
        }

        // Step 3: Sort each segment alphabetically
        Collections.sort(platinum);
        Collections.sort(gold);
        Collections.sort(silver);
        Collections.sort(bronze);
    }

    public void displayResults() {
        System.out.println();
        System.out.println("=".repeat(70));
        System.out.println("  CUSTOMER SEGMENTATION RESULTS");
        System.out.println("  Based on: Publisher Total Sales (in thousands of units)");
        System.out.println("=".repeat(70));

        System.out.println();
        System.out.println("  Segmentation Criteria:");
        System.out.println("  Platinum  ->  > 100,000 total sales");
        System.out.println("  Gold      ->  50,000 - 100,000 total sales");
        System.out.println("  Silver    ->  10,000 - 49,999 total sales");
        System.out.println("  Bronze    ->  < 10,000 total sales");

        System.out.println();
        System.out.println("=".repeat(70));
        System.out.println("  SEGMENT SUMMARY");
        System.out.println("=".repeat(70));
        System.out.printf("  %-12s : %d publishers%n", "Platinum", platinum.size());
        System.out.printf("  %-12s : %d publishers%n", "Gold",     gold.size());
        System.out.printf("  %-12s : %d publishers%n", "Silver",   silver.size());
        System.out.printf("  %-12s : %d publishers%n", "Bronze",   bronze.size());
        System.out.println("  " + "-".repeat(35));
        System.out.printf("  %-12s : %d publishers%n", "TOTAL",
                          platinum.size() + gold.size() + silver.size() + bronze.size());

        printSegment("PLATINUM", platinum, "> 100,000");
        printSegment("GOLD",     gold,     "50,000 - 100,000");
        printSegment("SILVER",   silver,   "10,000 - 49,999");
        printSegment("BRONZE",   bronze,   "< 10,000");

        System.out.println();
        System.out.println("=".repeat(70));
        System.out.println("  END OF REPORT");
        System.out.println("=".repeat(70));
    }

    private void printSegment(String label, List<String> publishers, String range) {
        System.out.println();
        System.out.println("=".repeat(70));
        System.out.println("  [" + label + "] - Sales Range: " + range);
        System.out.println("  Total Publishers: " + publishers.size());
        System.out.println("=".repeat(70));

        if (publishers.isEmpty()) {
            System.out.println("  No publishers in this segment.");
            return;
        }

        int count = 1;
        for (String publisher : publishers) {
            double sales = publisherSales.get(publisher);
            System.out.printf("  %3d. %-40s  Sales: %,.2f%n",
                              count++, publisher, sales);
        }
    }
}

File: MIDTERM_LAB/JS/CustomerSegmentation.js
// CustomerSegmentation.js
// Segments publishers by total sales into Platinum, Gold, Silver, Bronze

const PLATINUM_MIN = 100000;
const GOLD_MIN     = 50000;
const SILVER_MIN   = 10000;

function analyze(games) {
    // Step 1: Aggregate total sales per publisher
    const publisherSales = {};

    for (const game of games) {
        const publisher = game.publisher;
        const sales     = game.totalSales;

        if (publisher !== 'N/A' && sales > 0) {
            if (publisherSales[publisher]) {
                publisherSales[publisher] += sales;
            } else {
                publisherSales[publisher] = sales;
            }
        }
    }

    // Step 2: Segment each publisher
    const platinum = [];
    const gold     = [];
    const silver   = [];
    const bronze   = [];

    for (const publisher in publisherSales) {
        const sales = publisherSales[publisher];

        if (sales > PLATINUM_MIN) {
            platinum.push({ name: publisher, sales });
        } else if (sales >= GOLD_MIN) {
            gold.push({ name: publisher, sales });
        } else if (sales >= SILVER_MIN) {
            silver.push({ name: publisher, sales });
        } else {
            bronze.push({ name: publisher, sales });
        }
    }

    // Step 3: Sort each segment by sales descending
    const sortBySales = (a, b) => b.sales - a.sales;
    platinum.sort(sortBySales);
    gold.sort(sortBySales);
    silver.sort(sortBySales);
    bronze.sort(sortBySales);

    return { platinum, gold, silver, bronze };
}

function displayResults(segments) {
    const { platinum, gold, silver, bronze } = segments;

    console.log();
    console.log('='.repeat(70));
    console.log('  CUSTOMER SEGMENTATION RESULTS');
    console.log('  Based on: Publisher Total Sales (in thousands of units)');
    console.log('='.repeat(70));

    console.log();
    console.log('  Segmentation Criteria:');
    console.log('  Platinum  ->  > 100,000 total sales');
    console.log('  Gold      ->  50,000 - 100,000 total sales');
    console.log('  Silver    ->  10,000 - 49,999 total sales');
    console.log('  Bronze    ->  < 10,000 total sales');

    console.log();
    console.log('='.repeat(70));
    console.log('  SEGMENT SUMMARY');
    console.log('='.repeat(70));
    console.log('  Platinum  : ' + platinum.length + ' publishers');
    console.log('  Gold      : ' + gold.length     + ' publishers');
    console.log('  Silver    : ' + silver.length   + ' publishers');
    console.log('  Bronze    : ' + bronze.length   + ' publishers');
    console.log('  ' + '-'.repeat(35));
    console.log('  TOTAL     : ' + (platinum.length + gold.length + silver.length + bronze.length) + ' publishers');

    printSegment('PLATINUM', platinum, '> 100,000');
    printSegment('GOLD',     gold,     '50,000 - 100,000');
    printSegment('SILVER',   silver,   '10,000 - 49,999');
    printSegment('BRONZE',   bronze,   '< 10,000');

    console.log();
    console.log('='.repeat(70));
    console.log('  END OF REPORT');
    console.log('='.repeat(70));
}

function printSegment(label, publishers, range) {
    console.log();
    console.log('='.repeat(70));
    console.log('  [' + label + '] - Sales Range: ' + range);
    console.log('  Total Publishers: ' + publishers.length);
    console.log('='.repeat(70));

    if (publishers.length === 0) {
        console.log('  No publishers in this segment.');
        return;
    }

    publishers.forEach((p, i) => {
        const num   = String(i + 1).padStart(3, ' ');
        const name  = p.name.padEnd(40, ' ');
        const sales = p.sales.toFixed(2);
        console.log('  ' + num + '. ' + name + '  Sales: ' + sales);
    });
}

module.exports = { analyze, displayResults };